core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantExecutionStats.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import io.airlift.log.Logger;
+import io.airlift.stats.TimeStat;
+import io.trino.execution.ExecutionFailureInfo;
+import io.trino.execution.TaskInfo;
+import io.trino.execution.TaskState;
+import io.trino.operator.TaskStats;
+import io.trino.spi.ErrorCode;
+import io.trino.spi.ErrorType;
+import io.trino.spi.TrinoException;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import java.util.Optional;
+
+import static io.trino.spi.ErrorType.INTERNAL_ERROR;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.util.Failures.toFailure;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class FaultTolerantExecutionStats
+{
+    private static final Logger log = Logger.get(FaultTolerantExecutionStats.class);
+
+    private final ExecutionStats finishedTasks = new ExecutionStats();
+    private final FailedTasksStats failedTasks = new FailedTasksStats();
+
+    public void update(TaskInfo info)
+    {
+        TaskState state = info.getTaskStatus().getState();
+        switch (state) {
+            case FINISHED:
+                finishedTasks.update(info.getStats());
+                break;
+            case FAILED: {
+                failedTasks.update(info);
+                break;
+            }
+            case CANCELED:
+            case ABORTED:
+                // ignore cancelled and aborted tasks
+                break;
+            case PLANNED:
+            case RUNNING:
+            case FLUSHING:
+            default:
+                log.error("Unexpected task state: %s", state);
+        }
+    }
+
+    @Managed
+    @Nested
+    public ExecutionStats getFinishedTasks()
+    {
+        return finishedTasks;
+    }
+
+    @Managed
+    @Nested
+    public FailedTasksStats getFailedTasks()
+    {
+        return failedTasks;
+    }
+
+    public static class ExecutionStats
+    {
+        private final TimeStat elapsedTime = new TimeStat(MILLISECONDS);
+        private final TimeStat scheduledTime = new TimeStat(MILLISECONDS);
+        private final TimeStat cpuTime = new TimeStat(MILLISECONDS);
+        private final TimeStat inputBlockedTime = new TimeStat(MILLISECONDS);
+        private final TimeStat outputBlockedTime = new TimeStat(MILLISECONDS);
```

Comment:
Should we have memory and network stats here too?
It feels not costly to add them and then we can decide which are the most important for us for tracking.

Reply:
I'm not sure if there's a reliable metric for network as the network traffic can occur at many different levels (connector / exchanges / coordinator-to-worker communication). Though it certainly feels like it would make sense to record peak memory utilization. Let me add it.